src/player/player.cpp

```diff
@@ -22,51 +20,62 @@
 #include "../errors.hpp"
 #include "../io/io_response.hpp"
 #include "../messages.h"
-#include "player_state.hpp"
 #include "player.hpp"
 
 const std::vector<std::string> Player::FEATURES{ "End", "FileLoad", "PlayStop",
 	                                         "Seek", "TimeReport" };
 
-Player::Player(const ResponseSink *end_sink, PlayerFile &file,
-               PlayerPosition &position, PlayerState &state)
-    : file(file), position(position), state(state), end_sink(end_sink)
+Player::Player(AudioSystem &audio)
+    : audio(audio), file(audio.Null()), is_running(true), sink(nullptr)
 {
 }
 
+void Player::SetSink(ResponseSink &sink)
+{
+	this->sink = &sink;
+}
+
 bool Player::Update()
 {
-	auto as = this->file.Update();
+	assert(this->file != nullptr);
+	auto as = this->file->Update();
+
 	if (as == Audio::State::AT_END) this->End();
 	if (as == Audio::State::PLAYING) {
 		// Since the audio is currently playing, the position may have
 		// advanced since last update.  So we need to update it.
-		this->position.Update(this->file.Position());
+		this->file->Emit({ Response::Code::TIME }, this->sink);
```

Comment:
nou

On a slightly related note, I'm tempted to file a new PRQ afterwards changing `Emit` to take precisely one response code.  This'll cause some calls to `Emit` to expand into two or three calls, but simplify `Emit` and also make some of these items moot :grinning: